kv/dist_sender.go

```diff
@@ -486,44 +496,51 @@ func (ds *DistSender) sendRPC(trace *tracer.Trace, rangeID proto.RangeID, replic
 // descriptors associated with it. First, the range descriptor for
 // call.Args.Key is looked up. If call.Args.EndKey exceeds that of the
 // returned descriptor, the next descriptor is obtained as well.
-func (ds *DistSender) getDescriptors(call proto.Call) (*proto.RangeDescriptor, *proto.RangeDescriptor, error) {
-	// If this is a PushTxn, set ignoreIntents option as
+// The returned closure is to be called if the main range descriptor
+// is discovered to be stale.
+// TODO(tschottdorf): isReverse is deducible from the call, but it's
+// also awkward to re-scan the whole call all the time.
+func (ds *DistSender) getDescriptors(call proto.Call, isReverse bool) (*proto.RangeDescriptor, *proto.RangeDescriptor, func(), error) {
+	// If the call contains a PushTxn, set ignoreIntents option as
 	// necessary. This prevents a potential infinite loop; see the
 	// comments in proto.RangeLookupRequest.
 	options := lookupOptions{}
-	if pushArgs, ok := call.Args.(*proto.PushTxnRequest); ok {
-		options.ignoreIntents = pushArgs.RangeLookup
+	if arg, ok := proto.GetArg(call.Args, proto.PushTxn); ok {
+		options.ignoreIntents = arg.(*proto.PushTxnRequest).RangeLookup
 	}
 
 	var desc *proto.RangeDescriptor
 	var err error
-	_, isReverseScan := call.Args.(*proto.ReverseScanRequest)
-	if !isReverseScan {
-		desc, err = ds.rangeCache.LookupRangeDescriptor(call.Args.Header().Key, options)
+	header := call.Args.Header()
+	var descKey proto.Key
+	if !isReverse {
+		descKey = header.Key
+		desc, err = ds.rangeCache.LookupRangeDescriptor(descKey, options)
 	} else {
 		options.useReverseScan = true
-		desc, err = ds.rangeCache.LookupRangeDescriptor(call.Args.Header().EndKey, options)
+		descKey = header.EndKey
+		desc, err = ds.rangeCache.LookupRangeDescriptor(descKey, options)
 	}
 
 	if err != nil {
-		return nil, nil, err
+		return nil, nil, nil, err
 	}
 
 	// Checks whether need to get next range descriptor. If so, returns true
 	// and the key to look up, depending on whether we're in reverse mode.
-	needAnother := func(desc *proto.RangeDescriptor, isReverseScan bool) (proto.Key, bool) {
-		if isReverseScan {
-			return desc.StartKey, call.Args.Header().Key.Less(desc.StartKey)
+	needAnother := func(desc *proto.RangeDescriptor, isReverse bool) (proto.Key, bool) {
+		if isReverse {
+			return desc.StartKey, keys.KeyAddress(call.Args.Header().Key).Less(keys.KeyAddress(desc.StartKey))
```

Comment:
There should never be any reason to call `keys.KeyAddress` on a descriptor start or end key.